File: hello-csharp/src/Client/Core/DownloadManager.cs
using System;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;

namespace AudioFileTransfer.Client.Core;

/// <summary>
/// Download manager for handling file downloads
/// </summary>
public static class DownloadManager
{
    public static async Task DownloadAsync(WebSocketClient ws, string streamId, string outputPath, long fileSize)
    {
        // Ensure output directory exists
        string? directory = Path.GetDirectoryName(outputPath);
        if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
        {
            Directory.CreateDirectory(directory);
        }

        // Delete output file if it exists
        if (File.Exists(outputPath))
        {
            File.Delete(outputPath);
        }

        long offset = 0;
        long lastProgress = 0;

        while (offset < fileSize)
        {
            int length = ChunkManager.CalculateChunkSize(fileSize, offset);
            
            // Send GET message
            var getMessage = new ControlMessage
            {
                Type = "GET",
                StreamId = streamId,
                Offset = offset,
                Length = length
            };
            string getJson = JsonSerializer.Serialize(getMessage);
            await ws.SendTextAsync(getJson);
            Logger.Debug($"Requesting chunk at offset {offset}, length {length}");

            // Receive binary data
            byte[] data = await ws.ReceiveBinaryAsync();
            Logger.Debug($"Received {data.Length} bytes of data");

            // Write to file
            await FileManager.WriteChunkAsync(outputPath, data);

            offset += data.Length;

            // Report progress at 25%, 50%, 75%, 100%
            long progress = (offset * 100) / fileSize;
            if (progress >= lastProgress + 25 || offset == fileSize)
            {
                Logger.Info($"Download progress: {offset}/{fileSize} bytes ({progress}%)");
                lastProgress = progress;
            }
        }

        Logger.Info($"Download completed: {outputPath}");
    }
}

File: hello-nodejs/src/client/core/DownloadManager.js
/**
 * Download manager for receiving files from the server
 */

import { FileManager, CHUNK_SIZE } from "./FileManager.js";

export class DownloadManager {
  constructor(webSocketClient) {
    this.ws = webSocketClient;
    this.fileManager = new FileManager();
  }

  /**
   * Download a file from the server.
   *
   * @param {string} streamId - Stream ID to download
   * @param {string} outputPath - Path to save the file
   * @param {number} fileSize - Expected file size
   * @param {Function} progressCallback - Optional progress callback
   */
  async download(streamId, outputPath, fileSize, progressCallback = null) {
    let offset = 0;
    let bytesReceived = 0;
    let isFirstChunk = true;

    while (offset < fileSize) {
      const remainingBytes = fileSize - offset;
      const chunkSize = Math.min(CHUNK_SIZE, remainingBytes);

      // Send GET message
      await this.ws.sendControlMessage({
        type: "GET",
        streamId,
        offset,
        length: chunkSize,
      });

      // Receive binary data
      const data = await this.ws.receiveBinary();

      if (data.length === 0) {
        throw new Error(`No data received for offset ${offset}`);
      }

      // Write to file
      await this.fileManager.writeChunk(outputPath, data, !isFirstChunk);

      isFirstChunk = false;
      offset += data.length;
      bytesReceived += data.length;

      // Report progress
      if (progressCallback) {
        const progress = Math.floor((bytesReceived * 100) / fileSize);
        progressCallback(bytesReceived, fileSize, progress);
      }
    }
  }
}
